registry/memory/memory.go

```diff
@@ -74,91 +118,152 @@ func (m *Registry) Init(opts ...registry.Option) error {
+	}
+
+	// refresh TTL and timestamp
+	for _, n := range s.Nodes {
+		log.Debugf("Updated registration for service: %s, version: %s", s.Name, s.Version)
+		m.records[s.Name][s.Version].Nodes[n.Id].TTL = options.TTL
+		m.records[s.Name][s.Version].Nodes[n.Id].LastSeen = time.Now()
+	}
+
+	return nil
 }
 
-func (m *Registry) Register(s *registry.Service, opts ...registry.RegisterOption) error {
+func (m *Registry) Deregister(s *registry.Service) error {
 	m.Lock()
 	defer m.Unlock()
 
-	if service, ok := m.Services[s.Name]; !ok {
-		m.Services[s.Name] = []*registry.Service{s}
-		go m.sendEvent(&registry.Result{Action: "update", Service: s})
-	} else {
-		svcCount := len(service)
-		svcNodeCounts := make(map[string]map[string]int)
-		for _, s := range service {
-			if _, ok := svcNodeCounts[s.Name]; !ok {
-				svcNodeCounts[s.Name] = make(map[string]int)
+	if _, ok := m.records[s.Name]; ok {
+		if _, ok := m.records[s.Name][s.Version]; ok {
+			for _, n := range s.Nodes {
+				if _, ok := m.records[s.Name][s.Version].Nodes[n.Id]; ok {
+					log.Debugf("Registry removed node from service: %s, version: %s", s.Name, s.Version)
+					delete(m.records[s.Name][s.Version].Nodes, n.Id)
+				}
 			}
-			if _, ok := svcNodeCounts[s.Name][s.Version]; !ok {
-				svcNodeCounts[s.Name][s.Version] = len(s.Nodes)
+			if len(m.records[s.Name][s.Version].Nodes) == 0 {
+				delete(m.records[s.Name], s.Version)
+				log.Debugf("Registry removed service: %s, version: %s", s.Name, s.Version)
 			}
 		}
-		// if merged count and original service counts changed we added new version of the service
-		merged := registry.Merge(service, []*registry.Service{s})
-		if len(merged) != svcCount {
-			m.Services[s.Name] = merged
-			go m.sendEvent(&registry.Result{Action: "update", Service: s})
-			return nil
-		}
-		// if the node count for a particular service has changed we added a new node to the service
-		for _, s := range merged {
-			if len(s.Nodes) != svcNodeCounts[s.Name][s.Version] {
-				m.Services[s.Name] = merged
-				go m.sendEvent(&registry.Result{Action: "update", Service: s})
-				return nil
-			}
+		if len(m.records[s.Name]) == 0 {
+			delete(m.records, s.Name)
+			log.Debugf("Registry removed service: %s", s.Name)
 		}
+		go m.sendEvent(&registry.Result{Action: "delete", Service: s})
 	}
 
 	return nil
 }
 
-func (m *Registry) Deregister(s *registry.Service) error {
-	m.Lock()
-	defer m.Unlock()
+func (m *Registry) GetService(name string) ([]*registry.Service, error) {
+	m.RLock()
+	defer m.RUnlock()
 
-	if service, ok := m.Services[s.Name]; ok {
-		go m.sendEvent(&registry.Result{Action: "delete", Service: s})
-		if service := registry.Remove(service, []*registry.Service{s}); len(service) == 0 {
-			delete(m.Services, s.Name)
-		} else {
-			m.Services[s.Name] = service
+	records, ok := m.records[name]
+	if !ok {
+		return nil, registry.ErrNotFound
+	}
+
+	services := make([]*registry.Service, len(m.records[name]))
+	i := 0
+	for _, record := range records {
```

Comment:
Should you not just be able to do `for i, record` and get the index integer for free rather than incrementing a counter?

Reply:
Nope, because `records` is a map

`m.records` is your favourite data structure: map of maps; `records` here is a particular instance of the map of maps.